source/common/config/datasource.h

```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
  */
 absl::optional<std::string> getPath(const envoy::api::v2::core::DataSource& source);
 
+/**
+ * Callback for async data source.
+ */
+using AsyncDataSourceCb = std::function<void(std::string)>;
+
+class LocalAsyncDataProvider;
+using LocalAsyncDataProviderPtr = std::unique_ptr<LocalAsyncDataProvider>;
+
+class RemoteAsyncDataProvider;
+using RemoteAsyncDataProviderPtr = std::unique_ptr<RemoteAsyncDataProvider>;
+
+class LocalAsyncDataProvider {
+public:
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
```

Comment:
Warning: Take care. There is ongoing efforts on making Target remember `ready()` invoked before InitManager invokes `initialize()`. By then this `ready()` may not act as you expected